### ISSUE TITLE
Tensorflow hang when specify 'nccl/xring' as all reduce alg 

### ISSUE THREAD

**Issue by Keepmoving-ZXY** [tensorflow/tensorflow#26132]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:No
- TensorFlow installed from (source or binary):binary
- TensorFlow version (use command below):1.13.0-rc0
- Python version:2.7.12
- Bazel version (if compiling from source):0.20.0
- GCC/Compiler version (if compiling from source):5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.11)
- CUDA/cuDNN version:CUDA version:9.0.176 and cuDNN version:7.2.1.38
- GPU model and memory:Tesla V100 , 32480MiB

**Describe the current behavior**
I train NeuMF model in distribute environment, and I find that TensorFlow hang all the time when training finish.   
Below is some detail about distributed training:
Model : `NeuMF`
Dataset : `ml-1m`
Num_worker : `2`
Num_gpu_each_worker: `4`
Strategy : `tf.contrib.distribute.MirrorStrategy` 
Reduce_alg : replace `pscpu/pscpu` with `nccl/xring`
Nccl version : `2.3.7-1+cuda9.0`

The distributed training run 120 steps then tensorflow quit when use default all reduce alg `pscpu/pscpu` of `MirrorStrategy`. And when replace default all reduce alg with `nccl/xring` , distributed training run 120 steps, but TensorFlow hang all the time. Below is screen output of two conditions:  
log when normal exit:
```
I0226 10:04:21.481378 140388122244864 basic_session_run_hooks.py:247] loss = 0.3570031, step = 120 (1.436 sec)
I0226 10:04:21.483242 140388122244864 basic_session_run_hooks.py:680] global_step/sec: 6.96639
I0226 10:04:22.065148 140388122244864 util.py:168] Finalize strategy.
I0226 10:04:22.066411 140388122244864 basic_session_run_hooks.py:594] Saving checkpoints for 125 into /tmp/ncf_model/model.ckpt.
I0226 10:04:22.903732 140388122244864 estimator.py:359] Loss for final step: 0.35619634.
I0226 10:04:22.906255 140388122244864 data_preprocessing.py:391] Shutting down train data creation subprocess.
```
log when hang:
```
I0226 10:13:19.793405 140644590810880 basic_session_run_hooks.py:247] loss = 0.360583, step = 100 (1.755 sec)
I0226 10:13:19.795331 140644590810880 basic_session_run_hooks.py:680] global_step/sec: 5.70165
I0226 10:13:21.573970 140644590810880 basic_session_run_hooks.py:247] loss = 0.35865137, step = 110 (1.781 sec)
I0226 10:13:21.576019 140644590810880 basic_session_run_hooks.py:680] global_step/sec: 5.616
I0226 10:13:23.344115 140644590810880 basic_session_run_hooks.py:247] loss = 0.35676146, step = 120 (1.770 sec)
I0226 10:13:23.347656 140644590810880 basic_session_run_hooks.py:680] global_step/sec: 5.64505
``` 
Compare two log,  TensorFlow do not continue to save checkpoints and output final loss when use `nccl/xring` all reduce alg. How to solve this problem, thanks.

**Code to reproduce the issue**
I get NeuMF model source code from official models repo, and original code only support local training with estimator, and I add below code to support distributed training:  
```
distribution = tf.contrib.distribute.MirroredStrategy(
    num_gpus_per_worker=num_gpus)
    
run_config = tf.estimator.RunConfig(
    model_dir=model_dir,
    log_step_count_steps=10,
    train_distribute=distribution,                
    eval_distribute=distribution)
    
 estimator = tf.estimator.Estimator(
    model_fn=model_fn,
    model_dir=model_dir,
    config=run_config, 
    params=params)
    
 train_spec = tf.estimator.TrainSpec(
    train_input_fn,max_steps=5000,hooks=train_hooks)
 eval_spec = tf.estimator.EvalSpec(eval_input_fn,steps=100)
 tf.estimator.train_and_evaluate(estimator,train_spec,eval_spec)
```

The `cross device ops` implement of `MirrorStrategy`  is `MultiWorkerAllReduce`  which is defined in `tensorflow/python/distribute/cross_device_ops.py`, I change value of `all_reduce_alg`  parameter of `__init__`  of class `MultiworkerAllReduce` from `pscpu/pscpu` to `nccl/xring`.
  
**Other info / logs**
When hang, stack of two worker is:
```
#0  0x00007f990858ba13 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
#1  0x00007f988f250d98 in pollset_work () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#2  0x00007f988f270dcf in cq_pluck(grpc_completion_queue*, void*, gpr_timespec, void*) () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#3  0x00007f988f27119b in grpc_completion_queue_pluck () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#4  0x00007f988f1dc898 in grpc::CoreCodegen::grpc_completion_queue_pluck(grpc_completion_queue*, void*, gpr_timespec, void*) ()
   from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#5  0x00007f988882ad02 in grpc::CompletionQueue::Pluck(grpc::internal::CompletionQueueTag*) () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#6  0x00007f988882f056 in grpc::internal::BlockingUnaryCallImpl<tensorflow::RunStepRequest, tensorflow::RunStepResponse>::BlockingUnaryCallImpl(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc::ClientContext*, tensorflow::RunStepRequest const&, tensorflow::RunStepResponse*) () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#7  0x00007f988882f26d in tensorflow::grpc::MasterService::Stub::RunStep(grpc::ClientContext*, tensorflow::RunStepRequest const&, tensorflow::RunStepResponse*) ()
   from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#8  0x00007f98886b3f6d in tensorflow::GrpcRemoteMaster::RunStep(tensorflow::CallOptions*, tensorflow::RunStepRequestWrapper*, tensorflow::MutableRunStepResponseWrapper*) ()
   from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#9  0x00007f98886adfc9 in tensorflow::GrpcSession::RunProto(tensorflow::CallOptions*, tensorflow::MutableRunStepRequestWrapper*, tensorflow::MutableRunStepResponseWrapper*) ()
   from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#10 0x00007f98886aec52 in tensorflow::GrpcSession::RunHelper(tensorflow::RunOptions const&, std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tensorflow::Tensor>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tensorflow::Tensor> > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, tensorflow::RunMetadata*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
   from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#11 0x00007f98886af3e0 in tensorflow::GrpcSession::Run(tensorflow::RunOptions const&, std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tensorflow::Tensor>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tensorflow::Tensor> > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, tensorflow::RunMetadata*) () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#12 0x00007f9888693f42 in tensorflow::SessionRef::Run(tensorflow::RunOptions const&, std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tensorflow::Tensor>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tensorflow::Tensor> > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, tensorflow::RunMetadata*) () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#13 0x00007f98888997a1 in TF_Run_Helper(tensorflow::Session*, char const*, TF_Buffer const*, std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tensorflow::Tensor>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tensorflow::Tensor> > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, TF_Tensor**, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, TF_Buffer*, TF_Status*) [clone .constprop.664] ()
   from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#14 0x00007f9888899f9e in TF_SessionRun () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#15 0x00007f988868fbc9 in tensorflow::TF_SessionRun_wrapper_helper(TF_Session*, char const*, TF_Buffer const*, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<_object*, std::allocator<_object*> > const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*, TF_Status*, std::vector<_object*, std::allocator<_object*> >*) () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#16 0x00007f988868fc62 in tensorflow::TF_SessionRun_wrapper(TF_Session*, TF_Buffer const*, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<_object*, std::allocator<_object*> > const&, std::vector<TF_Output, std::allocator<TF_Output> > const&, std::vector<TF_Operation*, std::allocator<TF_Operation*> > const&, TF_Buffer*, TF_Status*, std::vector<_object*, std::allocator<_object*> >*) ()
   from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#17 0x00007f988864a70a in _wrap_TF_SessionRun_wrapper () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#18 0x00000000004bc4aa in PyEval_EvalFrameEx ()
#19 0x00000000004b9b66 in PyEval_EvalCodeEx ()
#20 0x00000000004c1f56 in PyEval_EvalFrameEx ()
#21 0x00000000004b9b66 in PyEval_EvalCodeEx ()
#22 0x00000000004d5669 in ?? ()
#23 0x00000000004a587e in PyObject_Call ()
#24 0x00000000004be51e in PyEval_EvalFrameEx ()
---Type <return> to continue, or q <return> to quit---
```
```
90      pthread_join.c: No such file or directory.
bt
#0  0x00007f65d1baf98d in pthread_join (threadid=140057709098752, thread_return=0x0) at pthread_join.c:90
#1  0x00007f6540377b97 in std::thread::join() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x00007f654ef477a0 in tensorflow::(anonymous namespace)::StdThread::~StdThread() () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/../libtensorflow_framework.so
#3  0x00007f6551b6c032 in tensorflow::GrpcServer::Join() () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#4  0x00007f6551be4a8c in TF_ServerJoin () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#5  0x00007f655198d756 in _wrap_TF_ServerJoin () from /home/zxy/.local/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#6  0x00000000004bc4aa in PyEval_EvalFrameEx ()
#7  0x00000000004b9b66 in PyEval_EvalCodeEx ()
#8  0x00000000004c1f56 in PyEval_EvalFrameEx ()
#9  0x00000000004b9b66 in PyEval_EvalCodeEx ()
#10 0x00000000004c17c6 in PyEval_EvalFrameEx ()
#11 0x00000000004b9b66 in PyEval_EvalCodeEx ()
#12 0x00000000004c17c6 in PyEval_EvalFrameEx ()
#13 0x00000000004b9b66 in PyEval_EvalCodeEx ()
#14 0x00000000004c1f56 in PyEval_EvalFrameEx ()
#15 0x00000000004b9b66 in PyEval_EvalCodeEx ()
#16 0x00000000004c1f56 in PyEval_EvalFrameEx ()
#17 0x00000000004b9b66 in PyEval_EvalCodeEx ()
#18 0x00000000004c1f56 in PyEval_EvalFrameEx ()
#19 0x00000000004b9b66 in PyEval_EvalCodeEx ()
#20 0x00000000004c1f56 in PyEval_EvalFrameEx ()
#21 0x00000000004b9b66 in PyEval_EvalCodeEx ()
#22 0x00000000004c17c6 in PyEval_EvalFrameEx ()
#23 0x00000000004b9b66 in PyEval_EvalCodeEx ()
#24 0x00000000004eb69f in ?? ()
#25 0x00000000004e58f2 in PyRun_FileExFlags ()
#26 0x00000000004e41a6 in PyRun_SimpleFileExFlags ()
#27 0x00000000004938ce in Py_Main ()
#28 0x00007f65d17fd830 in __libc_start_main (main=0x493370 <main>, argc=30, argv=0x7ffdb8177748, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffdb8177738)
    at ../csu/libc-start.c:291
#29 0x0000000000493299 in _start ()
```

**Comment by yuefengz** [comment]
Could you try with `CollectiveAllReduceStrategy`?  We have very limited support for MirroredStrategy running in distributed fashion.

**Comment by Keepmoving-ZXY** [comment]
`CollectiveAllReduceStrategy` is ok, but I think nccl is the best for local reduction, so there is an idea that use nccl in a worker and other method based on network for reduction between workers and `nccl/xring` achieve this idea but has problem. So does the `CollectiveAllReduceStrategy` also achieve this idea, does `CollectiveAllReduceStrategy` also use nccl for reduction in a worker?

**Comment by yuefengz** [comment]
Yeah, we will enable nccl implementation shortly in `CollectiveAllReduceStrategy`.

**Comment by Keepmoving-ZXY** [comment]
ok, thank you.

**Comment by yuefengz** [comment]
You can now use `tf.distribute.MultiWorkerMirroredStrategy(communication=tf.distribute.experimental.CollectiveCommunication.NCCL)` to use nccl on multiple nodes.

**Comment by Keepmoving-ZXY** [comment]
thank you, I find time to have a try  and I find this problem seems be related to NCCL version in these day.
